### PR TITLE
Throw an exception on startup if GroupCommand overloads are registered

### LINKED ISSUE #910: CNext should throw on multiple GroupCommand overloads
# Summary
Registering `GroupCommand` twice is perfectly acceptable to CNext, but has undocumented(?) behavior. 

# Details
If you were to register a group command twice (say, to allow for an overload), CommandsNext registers the commands, but only the first command is accessible. The overload exists, and registers, but seems to be inaccessible in terms of invoking it. Even when supplied with extra arguments, the overload doesn't seem to fire. 

# Steps to reproduce
- Make group commands
- Register two commands with the `[GroupCommand]` attributes.
- Watch as CNext registers, but doesn't allow you to invoke the secondary overload.

# Notes

image

### PR DESCRIPTION
**Pull request by ExaInsanity**
Previously, the library would allow GroupCommand overloads at startup, but at runtime they would fail to execute, throwing an `ArgumentException` with the message "Could not find a suitable overload for the command."

This PR moves the check to startup command registration instead and alerts the user with a clear error message.

fixes #910

### CODE REVIEW

File: DSharpPlus.CommandsNext/CommandsNextExtension.cs [hunk 1]
```diff
@@ -489,8 +489,19 @@ private void RegisterCommands(Type t, CommandGroupBuilder? currentParent, IEnume
                         foreach (var chk in inheritedChecks)
                             groupBuilder.WithExecutionCheck(chk);
 
-                        foreach (var mi in ti.DeclaredMethods.Where(x => x.IsCommandCandidate(out _) && x.GetCustomAttribute<GroupCommandAttribute>() != null))
-                            groupBuilder.WithOverload(new CommandOverloadBuilder(mi));
+                        if (ti.DeclaredMethods.Any(x => x.IsCommandCandidate(out _) && x.GetCustomAttribute<GroupCommandAttribute>() != null))
+                        {
+                            var groupCommandOverloads =
+                                ti.DeclaredMethods.Where(x => x.IsCommandCandidate(out _) && x.GetCustomAttribute<GroupCommandAttribute>() != null);
```
Comment (VelvetThePanda) [inline DSharpPlus.CommandsNext/CommandsNextExtension.cs:495]: Feels a bit daft to not merge these into one, or just use `Count(Func<T, bool>)`

File: DSharpPlus.CommandsNext/CommandsNextExtension.cs [hunk 2]
```diff
@@ -489,8 +489,19 @@ private void RegisterCommands(Type t, CommandGroupBuilder? currentParent, IEnume
...
+
+                            // overloads don't work for group commands, so throw on startup instead of at runtime.
+                            if(groupCommandOverloads.Count() > 1)
+                            {
+                                throw new InvalidOverloadException("GroupCommands cannot be overloaded", groupCommandOverloads.First());
```
Comment (VelvetThePanda) [inline DSharpPlus.CommandsNext/CommandsNextExtension.cs:500]: `$"Commands marked with [{nameof(GroupCommandAttribute)}] cannot be overloaded"` may be more apt

**Review by VelvetThePanda — CHANGES_REQUESTED** [review]

File: DSharpPlus.CommandsNext/CommandsNextExtension.cs (hunk 1 above)
Comment (ExaInsanity) [reply DSharpPlus.CommandsNext/CommandsNextExtension.cs:495]: this is solely a more-or-less pointless optimization; to never invoke any of that logic if no group commands are present in the first place

File: DSharpPlus.CommandsNext/CommandsNextExtension.cs (hunk 2 above)
Comment (ExaInsanity) [reply DSharpPlus.CommandsNext/CommandsNextExtension.cs]: implemented

File: DSharpPlus.CommandsNext/CommandsNextExtension.cs (hunk 1 above)
Comment (VelvetThePanda) [reply DSharpPlus.CommandsNext/CommandsNextExtension.cs:495]: Yes, but I don't see how this invalidates my point considering that in experience `.Count` is faster then `.Where().Count()`, which is being done below 😄

Comment (VelvetThePanda) [reply DSharpPlus.CommandsNext/CommandsNextExtension.cs:495]: Where's also the fact that you can slightly save on allocations by not allocating two iterators.

Comment (ExaInsanity) [reply DSharpPlus.CommandsNext/CommandsNextExtension.cs:495]: i suppose so, though invoking a `.Count()` just for it to be zero in a lot of cases, likely most cases doesn't seem right either.

Comment (VelvetThePanda) [reply DSharpPlus.CommandsNext/CommandsNextExtension.cs:495]: You're forcing an enumeration in all cases, so trying to guard with `.Any()` isn't a possible solution. I think we should just bite the bullet and check count anyway

File: DSharpPlus.CommandsNext/CommandsNextExtension.cs [hunk 3]
```diff
@@ -489,19 +489,14 @@ private void RegisterCommands(Type t, CommandGroupBuilder? currentParent, IEnume
...
-                        {
-                            var groupCommandOverloads =
-                                ti.DeclaredMethods.Where(x => x.IsCommandCandidate(out _) && x.GetCustomAttribute<GroupCommandAttribute>() != null);
-
-                            // overloads don't work for group commands, so throw on startup instead of at runtime.
-                            if(groupCommandOverloads.Count() > 1)
-                            {
-                                throw new InvalidOverloadException($"Commands marked with [{nameof(GroupCommandAttribute)}] cannot be overloaded", groupCommandOverloads.First());
-                            }
+                        var gcan = ti.DeclaredMethods.Where(x => x.IsCommandCandidate(out _) && x.GetCustomAttribute<GroupCommandAttribute>() != null).ToArray();
```
Comment (VelvetThePanda) [inline DSharpPlus.CommandsNext/CommandsNextExtension.cs:492]: "So long as we don't name it gcan" - Exa

Meanwhile exa's commit:

Comment (ExaInsanity) [reply DSharpPlus.CommandsNext/CommandsNextExtension.cs]: i got distracted fixing syntax. name fixed too, now

**Comment by VelvetThePanda** [comment]
"remove emzi naming" smh, might as well remove Poland from the globe while you're at it

**Pull request merged**